### PR TITLE
Implement "More contents" on PlaylistActionBubble

### PR DESCRIPTION
**Pull request by sangwoo108**
When we found media items that hasn't been saved to Playlist, we reveal this menu item and give users a chance to save them.

Resolves https://github.com/brave/brave-browser/issues/31635

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:
* Goto https://video-samples.glitch.me/multiple-videos.html
* Add one videos from multiple videos - you should exclude at least one.
* Check "More contents.." appears and test.

### CODE REVIEW

**Comment by simonhong** [comment]
@sangwoo108 it would be great to have test steps for this with real world urls.

File: browser/ui/views/playlist/playlist_action_bubble_view.cc [hunk 1]
```diff
@@ -437,19 +496,22 @@ PlaylistActionBubbleView* PlaylistActionBubbleView::GetBubble() {
...
 
 PlaylistActionBubbleView::~PlaylistActionBubbleView() = default;
 
 void PlaylistActionBubbleView::WindowClosing() {
   BubbleDialogDelegateView::WindowClosing();
 
-  DCHECK_EQ(g_bubble, this);
-  g_bubble = nullptr;
+  // g_bubble could be different one when new bubble is showing up subsequently.
+  if (g_bubble == this) {
```
Comment (simonhong) [inline browser/ui/views/playlist/playlist_action_bubble_view.cc:512]: I'm confusing between here and https://github.com/brave/brave-core/pull/19290/files#diff-1172219cf20634416f0b31a88f83d802a3bfded69ef8fa30cabf96cda8de08fcR41

In ShowBubble(), it checks `g_bubble` is null by `DCHECK(!g_bubble);`
It means, we assume that there is no active bubble when showing new bubble.
But, I think this check(`g_bubble == this`) means another bubble could be already shown
before closing previous bubble.
To me both checking seems contradict. confusing.. :)

**Comment by sangwoo108** [comment]
> @sangwoo108 it would be great to have test steps for this with real world urls.

I guess you can test it with any youtube videos. Currently, youtube returns 3 videos, which is known issue.

* Open up youtube sites.
* Add some videos from multiple videos
* Check "More contents.." appears and test.

File: browser/ui/views/playlist/playlist_action_bubble_view.cc (hunk 1 above)
Comment (sangwoo108) [reply browser/ui/views/playlist/playlist_action_bubble_view.cc:518]: Yeah, it's quite confusing. As we're calling OnWindowClosing() twice, `g_bubble != this` is true on the second call.

>  Call sequence | g_bubble state
> 
>  ---
> 
>  Close callback
>     WindowClosing | |this| to null
>     ShowBubble    | |null| to new bubble
>  WindowClosing callback
>     WindowClosing | shouldn't be changed.

For clarity, extract some part of code into a separate method - WindowClosingImpl().

**Comment by simonhong** [comment]
> > @sangwoo108 it would be great to have test steps for this with real world urls.
> 
> I guess you can test it with any youtube videos. Currently, youtube returns 3 videos, which is known issue.
> 
> * Open up a youtube video.
> * Add some videos from multiple videos - you should exclude at least one.
> * Check "More contents.." appears and test.

Ok, I could see two items from https://www.youtube.com/watch?v=gL--5nNwcFs.
When I added all at once I got below. Bottom separator should be removed.
image

To remove all, clicked `Remove from ...` and removed from playlist.
When I added one, then I could see `More media..` item.
And I added last one. but bubble still shows `More media..` item when clicks icon.

**Comment by sangwoo108** [comment]
> And I added last one. but bubble still shows More media.. item when clicks icon.

Yeah, that seems to be another issue I mentioned

> Currently, youtube returns 3 videos, which is known issue.

Youtube returns 2 or 3 videos but they might have the same media sourc URL. In that case, PlaylistService doesn't add item actually. That seems to make "More contents..." keep showing. I think it'd be better to handle that in a separate PR.

**Comment by sangwoo108** [comment]
For test, you can try this sample site https://video-samples.glitch.me/multiple-videos.html

**Comment by simonhong** [comment]
> For test, you can try this sample site https://video-samples.glitch.me/multiple-videos.html

Hmm.. got this failure with this sample url.

1. Loaded this url
2. Click playlist action icon
3. Bubble shows several items and all are checked state
4. Unchecked one item among them
5. Add button and got below

```
[32544:11760:0717/154810.584:FATAL:playlist_action_bubble_view.cc(41)] Check failed: !g_bubble.
Backtrace:
        base::debug::CollectStackTrace [0x00007FFD479DAE12+18] (C:\PRJ\brave-browser\src\base\debug\stack_trace_win.cc:329)
        base::debug::StackTrace::StackTrace [0x00007FFD479C3412+18] (C:\PRJ\brave-browser\src\base\debug\stack_trace.cc:218)
        logging::LogMessage::~LogMessage [0x00007FFD478986A1+177] (C:\PRJ\brave-browser\src\base\logging.cc:730)
        logging::`anonymous namespace'::DCheckLogMessage::~DCheckLogMessage [0x00007FFD478756B2+50] (C:\PRJ\brave-browser\src\base\check.cc:85)
        logging::CheckError::~CheckError [0x00007FFD47875421+33] (C:\PRJ\brave-browser\src\base\check.cc:262)
        PlaylistActionBubbleView::ShowBubble [0x00007FFCC9361C66+2046] (C:\PRJ\brave-browser\src\brave\browser\ui\views\playlist\playlist_action_bubble_view.cc:479)
        PlaylistActionIconView::ShowPlaylistBubble [0x00007FFCC937E1D5+565] (C:\PRJ\brave-browser\src\brave\browser\ui\views\playlist\playlist_action_icon_view.cc:77)
        base::TaskAnnotator::RunTaskImpl [0x00007FFD479294C2+482] (C:\PRJ\brave-browser\src\base\task\common\task_annotator.cc:186)
        base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWorkImpl [0x00007FFD47960E22+1714] (C:\PRJ\brave-browser\src\base\task\sequence_manager\thread_controller_with_message_pump_impl.cc:486)
        base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWork [0x00007FFD479600B4+324] (C:\PRJ\brave-browser\src\base\task\sequence_manager\thread_controller_with_message_pump_impl.cc:360)
        base::MessagePumpForUI::DoRunLoop [0x00007FFD479E9CED+317] (C:\PRJ\brave-browser\src\base\message_loop\message_pump_win.cc:212)
        base::MessagePumpWin::Run [0x00007FFD479E8F40+128] (C:\PRJ\brave-browser\src\base\message_loop\message_pump_win.cc:78)
        base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::Run [0x00007FFD47962345+613] (C:\PRJ\brave-browser\src\base\task\sequence_manager\thread_controller_with_message_pump_impl.cc:654)
        base::RunLoop::Run [0x00007FFD478F3C9B+491] (C:\PRJ\brave-browser\src\base\run_loop.cc:136)
        content::BrowserMainLoop::RunMainMessageLoop [0x00007FFD03AE1B82+162] (C:\PRJ\brave-browser\src\content\browser\browser_main_loop.cc:1069)
        content::BrowserMainRunnerImpl::Run [0x00007FFD03AE489E+46] (C:\PRJ\brave-browser\src\content\browser\browser_main_runner_impl.cc:160)
        content::BrowserMain [0x00007FFD03ADE726+182] (C:\PRJ\brave-browser\src\content\browser\browser_main.cc:34)
        content::RunBrowserProcessMain [0x00007FFD04AF7E0D+341] (C:\PRJ\brave-browser\src\content\app\content_main_runner_impl.cc:707)
        content::ContentMainRunnerImpl::RunBrowser [0x00007FFD04AF956E+762] (C:\PRJ\brave-browser\src\content\app\content_main_runner_impl.cc:1284)
        content::ContentMainRunnerImpl::Run [0x00007FFD04AF9111+897] (C:\PRJ\brave-browser\src\content\app\content_main_runner_impl.cc:1142)
        content::RunContentProcess [0x00007FFD04AF74E7+1319] (C:\PRJ\brave-browser\src\content\app\content_main.cc:326)
        content::ContentMain [0x00007FFD04AF763F+111] (C:\PRJ\brave-browser\src\content\app\content_main.cc:343)
        ChromeMain [0x00007FFCC7D5134C+684] (C:\PRJ\brave-browser\src\chrome\app\chrome_main.cc:187)

```

**Comment by simonhong** [comment]
> > And I added last one. but bubble still shows More media.. item when clicks icon.
> 
> Yeah, that seems to be another issue I mentioned
> 
> > Currently, youtube returns 3 videos, which is known issue.
> 
> Youtube returns 2 or 3 videos but they might have the same media sourc URL. In that case, PlaylistService doesn't add item actually. That seems to make "More contents..." keep showing. I think it'd be better to handle that in a separate PR, as it should include changes in PlaylistTabHelper or something around that.

Yeah, it's separated issue.

**Comment by sangwoo108** [comment]
> Bubble shows 6 items although this page has two videos


That's weird, I can't reproduce it. 🤔

**Comment by simonhong** [comment]
Yeah, tested on win11.

2023년 7월 17일 (월) 17:12, Sangwoo Ko ***@***.***>님이 작성:

**Comment by sangwoo108** [comment]
Thanks for checking. Unfortunately not reproducible on Win 10 😂 I think I should apply the same mechanism to "AddSelected()".

**Review by simonhong — APPROVED** [review]
++ Although more entries are added to playlist but it seems it's another issue now.
Works nicely :)

**Comment by sangwoo108** [comment]
> ++ Although more entries are added to playlist but it seems it's another issue now.

Yeah, thanks for checking that! Will work on it asap.

**Comment by fallaciousreasoning** [comment]
@simonhong you were too speedy :laughing:

**Pull request merged**